File: tests/sample_repo/orders/service.py
from orders.models import Order
from orders.repository import OrderRepository
from payments.gateway import PaymentGateway


class DuplicateOrderError(Exception):
    pass


class OrderService:
    """Entry point for order creation. Prevents duplicates via idempotency_key."""

    def __init__(self, repo: OrderRepository, gateway: PaymentGateway):
        self.repo = repo
        self.gateway = gateway

    def create(self, customer_id: str, amount_cents: int, idempotency_key: str) -> Order:
        existing = self.repo.get(idempotency_key)
        if existing is not None:
            raise DuplicateOrderError(idempotency_key)

        order = Order(customer_id=customer_id, amount_cents=amount_cents)
        order.idempotency_key = idempotency_key
        self.repo.save(order)

        try:
            self.gateway.charge(customer_id, amount_cents, idempotency_key)
        except Exception:
            order.mark_failed()
            self.repo.save(order)
            raise

        order.mark_paid()
        self.repo.save(order)
        return order

File: tests/sample_ts_repo/src/service.ts
import { Order } from "./models";
import { OrderRepository } from "./repository";
import { chargeCustomer } from "./gateway";

export class DuplicateOrderError extends Error {}

export class OrderService {
  constructor(private repo: OrderRepository) {}

  create(customerId: string, amountCents: number, idempotencyKey: string): Order {
    const existing = this.repo.findByKey(idempotencyKey);
    if (existing !== undefined) {
      throw new DuplicateOrderError(idempotencyKey);
    }

    const order = new Order(customerId, amountCents, idempotencyKey);
    this.repo.save(order);

    try {
      chargeCustomer(customerId, amountCents, idempotencyKey);
    } catch (err) {
      order.markFailed();
      this.repo.save(order);
      throw err;
    }

    order.markPaid();
    this.repo.save(order);
    return order;
  }
}
